8267988: C2: assert(!addp->is_AddP() || addp->in(AddPNode::Base)->is_top() || addp->in(AddPNode::Base) == n->in(AddPNode::Base)) failed: Base pointers must match (addp 1301)

rwestrel opened the pull request:
The failure occurs because, in a loop body, two AddP nodes are
chained:
(AddPNode base (AddPNode base ..) ..)

One can be sunk out of loop but not the other which results in:
(AddPNode (CastPP base) (AddPNode base ..) ..)

And the 2 AddP nodes no longer have the same base.

The fix I propose is to make sure sinking nodes out of loop preserve
the invariant that 2 chained AddP nodes have the same base. It's not
straightfoward given the 2 AddP nodes are sunk one after the other and
when sinking the first one, it's hard to tell whether the second one
can be sunk as well.

With the fix:

- when the first AddP is sunk, no CastPP node is added to pin the AddP
  out of the loop so if sinking the second AddP fails, the 2 AddP
  nodes still have the same base.

- when the second AddP is sunk, a CastPP node is created and wired so
  both AddP base is updated to the new CastPP.

This change also addresses another unrelated minor issue. If, for
instance the loop has a null check, the loop body already includes a
CastPP. So if that CastPP is used by an AddP that is sunk, we end up
with:

(AddPNode (CastPP (CastPP...

Then code in ConstraintCastNode::dominating_cast(), causes the CastPP
added to pin the AddP out of loop to be replaced by the CastPP in the
loop so sinking the AddP has no effect. To prevent that, I tweaked the
way a ConstrainedCast is marked as carrying a dependency so
ConstraintCastNode::dominating_cast() doesn't trigger.


/cc hotspot-compiler

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267988](https://bugs.openjdk.java.net/browse/JDK-8267988): C2: assert(!addp->is_AddP() || addp->in(AddPNode::Base)->is_top() || addp->in(AddPNode::Base) == n->in(AddPNode::Base)) failed: Base pointers must match (addp 1301)


### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to 709809a9da711429ec5f29e730a080b1e2d0136a
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 709809a9da711429ec5f29e730a080b1e2d0136a
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4387/head:pull/4387` \
`$ git checkout pull/4387`

Update a local copy of the PR: \
`$ git checkout pull/4387` \
`$ git pull https://git.openjdk.java.net/jdk pull/4387/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4387`

View PR using the GUI difftool: \
`$ git pr show -t 4387`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4387.diff">https://git.openjdk.java.net/jdk/pull/4387.diff</a>

</details>

chhagedorn commented on src/hotspot/share/opto/castnode.hpp:
```diff
@@ -32,15 +32,21 @@
+  enum DependencyType {
+    RegularDependency, // if cast doesn't improve input type, cast can be removed
+    StrongDependency,  // leave cast in even if _type doesn't improve input type, can be replaced by stricter dominating cast if one exist
+    VeryStrongDependency // leave cast in unconditionally
```
How about directly using something like `UnconditionalDependency` instead of `VeryStrongDepedendency` to better distinguish it?

rwestrel replied:
Done. I agree the naming scheme I chose is not great but I haven't found anything better so far.

chhagedorn commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -1503,7 +1503,7 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
           register_new_node(x, x_ctrl);
 
-          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr()) {
+          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr() && (!x->is_AddP() || !x->in(AddPNode::Address)->is_AddP())) {
```
Can you add a comment here about this special case?

rwestrel replied:
Done.

chhagedorn reviewed (approved):
Looks reasonable to me!

rwestrel commented:
> Looks reasonable to me!

Thanks for reviewing this.

chhagedorn reviewed (approved):
Looks good, thanks for the changes!

vnkozlov reviewed (commented):
@chhagedorn please run mach5 testing for this fix.

vnkozlov commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -1512,23 +1515,32 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
+                if (x->is_AddP() && k == AddPNode::Base) {
+                  for (DUIterator_Fast imax, i = x->fast_outs(imax); i < imax; i++) {
+                    Node* u = x->fast_out(i);
+                    if (u->is_AddP()) {
```
I think we need to check that originally both AddP point to the same base.
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/escape.cpp#L2502

rwestrel replied:
> I think we need to check that originally both AddP point to the same base.

Right. I pushed an updated change with a fix for that.

chhagedorn commented:
> @chhagedorn please run mach5 testing for this fix.

I've started some testing on the state of this branch. Applying the changes to latest master needs merging and resolving conflicts first.

rwestrel commented:
> > @chhagedorn please run mach5 testing for this fix.
> 
> I've started some testing on the state of this branch. Applying the changes to latest master needs merging and resolving conflicts first.

I'll update the change later today.

rwestrel commented:
> I'll update the change later today.

Done.

chhagedorn reviewed (approved):
I submitted testing for it.

iwanowww commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -1503,19 +1503,32 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
+          // when the first one is sunk so if the second one is not, their bases remain the same. A CastPP of the base
+          // is only added once both AddP nodes are sunk (see special code for AddP after the cast is created).
+          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr() &&
+              !(x->is_AddP() && x->in(AddPNode::Address)->is_AddP() && x->in(AddPNode::Address)->in(AddPNode::Base) == x->in(AddPNode::Base))) {
```
I read the last check as "not a case of chained AddP with different bases". 
Then the comparison should be `!=`, shouldn't it?

rwestrel replied:
It's safe to add a cast if it's not a chained AddP with the same base otherwise there's a risk that adding a cast would make the base differ.

iwanowww replied:
Got it. Wasn't attentive enough when reading the description and the comment.

iwanowww commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -1503,19 +1503,32 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
               if (cast != NULL) {
                 register_new_node(cast, x_ctrl);
                 x->replace_edge(in, cast);
+                if (x->is_AddP() && k == AddPNode::Base) {
```
Does the code assume that "inner" node is processed first?
Otherwise, the "outer" node could be already processed and bases don't match anymore.

I'm asking because `PhaseIdealLoop::split_if_with_blocks()` (which calls `split_if_with_blocks_post()` and `try_sink_out_of_loop()` does process users first.

rwestrel replied:
The expectation is that uses are processed first.  The:
!(x->is_AddP() && x->in(AddPNode::Address)->is_AddP() && x->in(AddPNode::Address)->in(AddPNode::Base) == x->in(AddPNode::Base))
check guarantees the "outer" node doesn't change the base when it's processed. When the "inner" node is processed, bases for the inner and outer node are updated.

iwanowww replied:
A comment briefly explaining what's happening here may help as well.

iwanowww reviewed (approved):
Thanks for the clarifications, Roland.

Looks good.

A more general question: does it make sense to migrate away from pointer equality (`in(AddPNode::Base) == in(AddPNode::Address)->in(AddPNode::Base)`) to `Node::eqv_uncast()` (`in(AddPNode::Base)->eqv_uncast(in(AddPNode::Address)->in(AddPNode::Base))`) when checking the shape of address computation? If it is doable, it would avoid the whole class of related bugs.

chhagedorn commented:
Testing looked good!

rwestrel commented:
> A more general question: does it make sense to migrate away from pointer equality (`in(AddPNode::Base) == in(AddPNode::Address)->in(AddPNode::Base)`) to `Node::eqv_uncast()` (`in(AddPNode::Base)->eqv_uncast(in(AddPNode::Address)->in(AddPNode::Base))`) when checking the shape of address computation? If it is doable, it would avoid the whole class of related bugs.

My understanding is that this shouldn't happen. So an assert like: assert(in(AddPNode::Base) == in(AddPNode::Address)->in(AddPNode::Base) ||  !in(AddPNode::Base)->eqv_uncast(in(AddPNode::Address)->in(AddPNode::Base))?

iwanowww commented:
> My understanding is that this shouldn't happen. So an assert like: assert(in(AddPNode::Base) == in(AddPNode::Address)->in(AddPNode::Base) || !in(AddPNode::Base)->eqv_uncast(in(AddPNode::Address)->in(AddPNode::Base))?

Yes, that's the invariant currently in place. And you are fixing a bug which breaks it. 

My question was about would it take to remove the requirement that bases are the same and replace it with a more relaxed varaint for bases to be equivalent modulo cast nodes (in `Node::eqv_uncast()` sense).

Introducing the assert makes perfect sense.

rwestrel commented:
> My question was about would it take to remove the requirement that bases are the same and replace it with a more relaxed varaint for bases to be equivalent modulo cast nodes (in `Node::eqv_uncast()` sense).

I'm not sure. I suppose it would require changes to several places and also that some transformations might not trigger or trigger incorrectly. Some of those might be silent (c2 wouldn't crash) but lead to slightly poorer code. So it might prove hard to gain sufficient confidence that a fix is complete. That's why I would rather go with the local fix.

> Introducing the assert makes perfect sense.

I added an assert and reworked the comment.

This will have to be fixed in 17 but I suppose it's easier to finish the review here first.

iwanowww commented:
> That's why I would rather go with the local fix.

I'm perfectly fine with the proposed fix.

I'll think a bit more about `Node::eqv_uncast()` and file an RFE if it turns out to be a feasible thing to experiment with.
